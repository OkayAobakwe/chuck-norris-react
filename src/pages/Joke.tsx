import React, { useContext, useEffect, useState } from "react";
import { JokeContext } from "../context/JokeContext";
import axios from "axios"
import { API_URL } from "../utils/categoriesUrl";
import { Card, CardContent, Typography, Grid, Container } from "@material-ui/core";

export const Joke: React.FC = () => {
  const category = useContext(JokeContext)?.selectedCategory;
  const [joke, setJoke] = useState("");
  
  useEffect(() => {
    if(category !== ""){
      axios
      .get(`${API_URL}random?category=${category}`)
      .then((res) => setJoke(res?.data?.value))
      .catch((err) => {})
    }
  }, [category]);
  
  return (
    <>
      <Container>
        <Grid>
          <Grid>
            <Card>
              <CardContent>
                <Typography variant="h5" component="h2">
                  Chuck Norris Joke from {category}
                  <hr/>
                </Typography>
                <Typography>
                  {joke}
                </Typography>
              </CardContent>
            </Card>
          </Grid>
        </Grid>
      </Container>
    </>
  )
}