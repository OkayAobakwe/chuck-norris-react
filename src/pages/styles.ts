import { styled } from "@material-ui/core/styles";
import {
  Grid, 
  Select,
  Typography,
  AppBar,
} from "@material-ui/core";

export const OptionsGrid = styled(Grid)({
  display: "flex",
  justifyContent: "space-between",
  marginBotton: "30px",
});
export const SelectCategories = styled(Select)({
  justifySelf: "flex-start",
  width: "25%",
})
export const SearchGrid = styled(Grid)({
  justifySelf: "flex-end",
  alignItems: "baseline",
})
export const CenteredText = styled(Typography)({
  display: "flex",
  justifyContent: "center"
})
export const ChuckBar = styled(AppBar)({
  marginBottom: "15px",
})