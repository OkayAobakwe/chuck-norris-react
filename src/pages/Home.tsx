import React, { useContext, useState } from "react";
import { CategoriesContext } from "../context/CategoryListContext";
import { JokeContext } from "../context/JokeContext";
import { Joke } from "./Joke";
import { SearchJoke } from "../components/SearchJoke";
import axios from "axios";
import { API_URL } from "../utils/categoriesUrl";

import {
  TextField,
  Button,
  MenuItem,
  Select,
  Grid,
  Fade,
  Container,
  Card,
  CardContent,
  Typography,
  AppBar,
} from "@material-ui/core";
import { styled } from "@material-ui/core/styles";

const OptionsGrid = styled(Grid)({
  display: "flex",
  justifyContent: "space-between",
  marginBotton: "30px",
});
const SelectCategories = styled(Select)({
  justifySelf: "flex-start",
  width: "25%",
})
const SearchGrid = styled(Grid)({
  justifySelf: "flex-end",
  alignItems: "baseline",
})
const CenteredText = styled(Typography)({
  display: "flex",
  justifyContent: "center"
})

export const Home: React.FC = () => {
  const { categories } = useContext(CategoriesContext);
  const [selectedCategory, setSelectedCategory] = useState("");
  const [query, setQuery] = useState("");
  const [searchResults, setSearchResults] = useState<[]>();

  const handleChange = (e: React.ChangeEvent<{ value: unknown }>) => {
    setSelectedCategory(e.target.value as string);
    setQuery("")
  };
  const handleInput = (e: React.ChangeEvent<HTMLInputElement>) => {
    setQuery(e.currentTarget.value);
  };
  const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (query !== "") {
      axios
        .get(`${API_URL}search?query=${query}`)
        .then((res) => {
          setSearchResults(res?.data?.result);
          setSelectedCategory("");
        })
        .catch((err) => {});
    }
  };

  return (
    <JokeContext.Provider value={{ selectedCategory: selectedCategory }}>
      <Fade in={true}>
        <div>
          <AppBar position="relative">
            <Typography variant="h3" noWrap>
              Chuck Norris Jokes
            </Typography>
          </AppBar>
          
          <OptionsGrid>
            <SelectCategories value={selectedCategory} onChange={handleChange} placeholder="Categories">
              {categories?.map((category) => (
                <MenuItem value={category} key={category}>
                  {category}
                </MenuItem>
              ))}
            </SelectCategories>
            <SearchGrid>
              <TextField
                onChange={handleInput}
                id="standard-basic"
                label="Search"
              />
              <Button onClick={handleClick}>search</Button>
            </SearchGrid>
          </OptionsGrid>

          {selectedCategory === "" && query !== "" ? (
            <Container>
              <Grid>
                {searchResults?.map((result: { value: string; id: string }) => (
                  <Grid>
                    <Card>
                      <CardContent>
                        <Typography>
                          <SearchJoke
                            searchResult={result.value}
                            key={result.id}
                          />
                        </Typography>
                      </CardContent>
                    </Card>
                  </Grid>
                ))}
              </Grid>
            </Container>
          ) : selectedCategory !== "" ? (
            <Joke />
          ) : (
            <CenteredText variant="h6">Please search a keyword or select a category</CenteredText>
          )}
        </div>
      </Fade>
    </JokeContext.Provider>
  );
};
