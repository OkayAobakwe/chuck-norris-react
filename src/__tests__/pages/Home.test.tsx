import { Home } from "../../pages/Home";
import { render } from "@testing-library/react";

describe("Home Page", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = render(<Home />);
  });

  it("renders successfully", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
