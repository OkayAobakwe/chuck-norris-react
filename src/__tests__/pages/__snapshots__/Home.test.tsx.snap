// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page renders successfully 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      >
        <header
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionRelative MuiAppBar-colorPrimary MuiPaper-elevation4"
        >
          <h3
            class="MuiTypography-root MuiTypography-h3 MuiTypography-noWrap"
          >
            Chuck Norris Jokes
          </h3>
        </header>
        <div
          class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-1"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline WithStyles(ForwardRef(Select))-root-2"
          >
            <div
              aria-haspopup="listbox"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              role="button"
              tabindex="0"
            >
              <span>
                ​
              </span>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              placeholder="Categories"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
          <div
            class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-3"
          >
            <div
              class="MuiFormControl-root MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                data-shrink="false"
                for="standard-basic"
                id="standard-basic-label"
              >
                Search
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="standard-basic"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                search
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <h6
          class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-4 MuiTypography-h6"
        >
          Please search a keyword or select a category
        </h6>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionRelative MuiAppBar-colorPrimary MuiPaper-elevation4"
      >
        <h3
          class="MuiTypography-root MuiTypography-h3 MuiTypography-noWrap"
        >
          Chuck Norris Jokes
        </h3>
      </header>
      <div
        class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-1"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline WithStyles(ForwardRef(Select))-root-2"
        >
          <div
            aria-haspopup="listbox"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
            role="button"
            tabindex="0"
          >
            <span>
              ​
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            placeholder="Categories"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
        <div
          class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-3"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
              for="standard-basic"
              id="standard-basic-label"
            >
              Search
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                id="standard-basic"
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              search
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <h6
        class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-4 MuiTypography-h6"
      >
        Please search a keyword or select a category
      </h6>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
